api/agent/drivers/docker/docker.go

```diff
@@ -269,55 +269,98 @@ func dockerMsg(derr *docker.Error) string {
 
 // Run executes the docker container. If task runs, drivers.RunResult will be returned. If something fails outside the task (ie: Docker), it will return error.
 // The docker driver will attempt to cast the task to a Auther. If that succeeds, private image support is available. See the Auther interface for how to implement this.
-func (drv *DockerDriver) run(ctx context.Context, container string, task drivers.ContainerTask) (drivers.WaitResult, error) {
+func (drv *DockerDriver) run(ctx0 context.Context, container string, task drivers.ContainerTask) (drivers.WaitResult, error) {
 	timeout := task.Timeout()
 
-	var cancel context.CancelFunc
+	// we create and use the following contexts in this func
+	// ctx0        - parent context
+	// ctxForStats - new child context (with timeout if specified) - passed to collectStats and cancelled when the waiter completes
+	// ctx2        - new child context (with timeout if specified) - used for remaining calls and cancelled at end of func
```

Comment:
i believe most of these were created here simply due to the confusion made by me of leaving the `defer cancel()` and `go drv.collectStats` lines in after heavily modifying `Run`'s contract -- apologies. 

I believe we can simplify this greatly. I kinda wrote out a blurb but I think it's easier to illustrate.

```go
func (drv *DockerDriver) run(ctx context.Context, container string, task drivers.ContainerTask) (drivers.WaitResult, error) {
        mwOut, mwErr := task.Logger()

        waiter, err := drv.docker.AttachToContainerNonBlocking(ctx, docker.AttachToContainerOptions{
                Container: container, OutputStream: mwOut, ErrorStream: mwErr,
                Stream: true, Logs: true, Stdout: true, Stderr: true,
                Stdin: true, InputStream: task.Input()})
        if err != nil && ctx.Err() == nil {
                // ignore if ctx has errored, rewrite status lay below
                return nil, err
        }

        err = drv.startTask(ctx, container)
        if err != nil && ctx.Err() == nil {
                // if there's just a timeout making the docker calls, drv.wait below will rewrite it to timeout
                return nil, err
        }

        ctx, cancel := context.WithCancel(ctx) // to shut down collectStats once container exits
        go drv.collectStats(ctx, container, task)

        return &waitResult{
                container: container,
                waiter:    waiter,
                drv:       drv,
                cancel: cancel,
        }, nil
}

// implements drivers.WaitResult
type waitResult struct {
        container string
        waiter    docker.CloseWaiter
        drv       *DockerDriver
        cancel   func()
}

func (w *waitResult) Wait(ctx context.Context) (drivers.RunResult, error) {
        defer func() {
                w.waiter.Close()
                w.waiter.Wait() // make sure we gather all logs
                w.cancel()
        }()

        status, err := w.drv.wait(ctx, w.container)
        return &runResult{
                status: status,
                err:    err,
        }, nil
}
```

we should remove `Timeout()` from the `drivers.ContainerTask` interface imo, we are not using it really. we are effectively waiting for the task to finish or for the enclosing context to be up, at which point we're nuking the container. we're not handing the timeout off as a parameter to the container at all, the implementation in `run` left around is now just confusing. 

the only addition is that maybe putting `collectStats` below `startTask` is too late for quick [cold, hot is fine] tasks, in which case we can move it back up but in any enclosing context that calls `cookie.Run()` we just need to make sure we cancel the given context once the waiter returns. the only confusing bit is... well, the contract with `Run()` is such that a user _could_ call it, run off and not wait for it, and then `collectStats` would run forever. I think this is why the additional goroutine was added inside of `Run`. `Run`, now, seems to have a weird contract, but in any event, we are always waiting, so I think we can expect that (we could demand it, in the contract, i.e. 'Wait.Wait` must always be called or resources will be leaked; this seems wise to me anyway).

too many words I think, but anyway, lmk what you think / what this may be missing.

Reply:
That certainly looks simpler. Obvious you have some additional information that I didn't have, that two aspects of the existing implementation (which I was preserving in the new version) are not actually needed:
* honouring the timeout set in the `ContainerTask`, and 
* ensuring that the context passed to `AttachToContainerNonBlocking` and `startTask` is cancelled at the end of `Run`. 
I'm happy to take your advice and take the extra contexts out.

As you say above, your suggested new implementation relies on the caller calling `WaitResult`. I was trying to avoid this coupling. However if we're happy with this coupling then I agree we don't need to have `Run() ` spawn a goroutine to call `WaitResult`. This also avoids the need to make  `WaitResult` support multiple calls to `Wait`. 

Passing in the cancel function to `waitResult` is a nice way to avoid depending on the parent context being cancelled.

However I'm unsure whether we should be moving the call to `collectStats` to after `startTask`. As you suggest, this makes it more likely that the container will start and stop before we have had time to collect any stats.  In a few quick experiments I found that with a cold function that  performed a 300ms sleep, I observed one set of stats being received if `collectStats` was called before `startTask`, but not if it was called after it.

However It's not trivial to move it back to its original position before `startTask`. Can you clarify what you had in mind in this case? I think we would need to create a new context (so we could cancel it without cancelling the parent context), but we would need to ensure that it wasn't cancelled until the container had exited. However `go vet` would then complain that it was possible to return without cancelling the new context. 

(I think `vet` is being inconsistent here, since exactly the same issue applies if we put the call to `collectStats` after `startTask` and reuse the `ctx` variable. It was the need to avoid a vet error like this that led me into the more complex implementation I proposed.)

Reply:
> As you say above, your suggested new implementation relies on the caller calling WaitResult. I was trying to avoid this coupling. However if we're happy with this coupling then I agree we don't need to have Run() spawn a goroutine to call WaitResult. This also avoids the need to make WaitResult support multiple calls to Wait.

this seems a reasonable contract to me, +1. it seems like we already demand `Wait` be called, we just need to add that it may only be called once to the contract there. 

> However I'm unsure whether we should be moving the call to collectStats to after startTask. As you suggest, this makes it more likely that the container will start and stop before we have had time to collect any stats. In a few quick experiments I found that with a cold function that performed a 300ms sleep, I observed one set of stats being received if collectStats was called before startTask, but not if it was called after it.

would forcing collectStats to collect >= 1 data point fix this? I am unsure of plumbing, obviously we don't want to have to wait on this just to return the call result to the user. We are already suffering in that many functions run too fast to collect metrics (few ms), so this seems something we should fix anyway. 

> It was the need to avoid a vet error like this that led me into the more complex implementation I proposed.

if vet is going to be a nazi we can turn it off, it's nice that it enforces good practices robotically vs us having to explain to contributors some subjectively better way of doing things but meh, we can go back to the old fashioned way if it's a hindrance.